ibazel/lifecycle.go

```diff
@@ -0,0 +1,25 @@
+package main
+
+import (
+	"github.com/bazelbuild/bazel-watcher/third_party/bazel/master/src/main/protobuf"
+)
+
+// Lifecycle is an object that listens to the lifecycle events of iBazel and
+// behaves appropriately..
+type Lifecycle interface {
+	// TargetDecider takes a protobuf rule and performs setup if it matches the
+	// listener's expectations.
+	TargetDecider(rule *blaze_query.Rule)
+
+	// Setup is called once it is known that this lifesycle client is going to be
+	// used. You can call methods on the ibazel object in this context to set
+	// additional environment variables to be passed into the client or to
+	// retrigger action.
+	Setup()
+	// Cleanup is your opportunity to clean up open sockets or connections.
+	Cleanup()
+
+	// Before running an "event" where name = (build|test|run).
+	BeforeEvent(string)
+	AfterEvent(string)
```

Comment:
Seems like you should pass a bit more context. For example, if the live reload server wanted to make an optimization by rereading only changed files, it would need to know which build targets were affected by the build event. Or if I'm missing something maybe you could add a comment to explain how to hook into metadata like that

Reply:
In the particular case of the live reload server, there can only be one target being run at a time. As a result, in the particular case of live reload the answer to "what target/files should I reload" is always going to be all of them.

There may, however. be some special logic that is useful in a `ibazel test` scenario. Lemme think on that.